### PR TITLE
Add typesafe way of modifying schema

### PR DESCRIPTION
**Pull request by ghostbuster91**
This closes #275

### CODE REVIEW

File: core/src/main/scala-2.13+/sttp.tapir/ModifyMacroSupport.scala [hunk 1]
```diff
@@ -0,0 +1,17 @@
+package sttp.tapir
+
+import scala.annotation.compileTimeOnly
+import scala.collection.Factory
+
+trait ModifyMacroSupport extends ModifyMacroOptionSupport {
```
Comment (adamw) [inline core/src/main/scala-2.13+/sttp.tapir/ModifyMacroSupport.scala]: I think I'd put these into `sttp.tapir.internal`

File: core/src/main/scala-2.13+/sttp.tapir/ModifyMacroSupport.scala [hunk 2]
```diff
@@ -0,0 +1,17 @@
+package sttp.tapir
```
Comment (adamw) [inline core/src/main/scala-2.13+/sttp.tapir/ModifyMacroSupport.scala]: is it github or is this file in the `sttp.tapir` directory? (not `sttp/tapir`)?

File: core/src/main/scala/sttp/tapir/Schema.scala [hunk 3]
```diff
@@ -40,10 +41,12 @@ case class Schema[T](
 
   def show: String = s"schema is $schemaType${if (isOptional) " (optional)" else ""}"
 
-  def modifyUnsafe(fields: String*)(modify: Schema[_] => Schema[_]): Schema[T] = modifyAtPath(fields.toList, modify)
+  def modifyUnsafe[U](fields: String*)(modify: Schema[U] => Schema[U]): Schema[T] = modifyAtPath(fields.toList, modify)
```
Comment (adamw) [inline core/src/main/scala/sttp/tapir/Schema.scala]: do we need this `[U]`? It seems it only causes a type parameter ot be required at use-site

File: core/src/main/scala/sttp/tapir/package.scala [hunk 4]
```diff
@@ -2,7 +2,7 @@ package sttp
 
 import sttp.model.Part
 
-package object tapir extends Tapir {
+package object tapir extends Tapir with ModifyMacroSupport {
```
Comment (adamw) [inline core/src/main/scala/sttp/tapir/package.scala]: shouldn't `Tapir` extends `ModifyMacroSupport`?

File: core/src/main/scala-2.13+/sttp.tapir/ModifyMacroSupport.scala [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+
+trait ModifyMacroSupport extends ModifyMacroOptionSupport {
+  implicit def traversableModifyFunctor[F[_], A](
+      implicit fac: Factory[A, F[A]],
+      ev: F[A] => Iterable[A]
+  ): ModifyFunctor[F, A] =
+    new ModifyFunctor[F, A] {}
+
+  implicit class ModifyEachMap[F[_, _], K, T](t: F[K, T])(implicit fac: Factory[(K, T), F[K, T]]) {
+    @compileTimeOnly(canOnlyBeUsedInsideIgnore("each"))
```
Comment (adamw) [inline core/src/main/scala-2.13+/sttp.tapir/ModifyMacroSupport.scala]: insideIgnore? ;)

File: core/src/main/scala-common/sttp.tapir/ModifyMacroOptionSupport.scala [hunk 6]
```diff
@@ -0,0 +1,21 @@
+package sttp.tapir
+
+import scala.annotation.compileTimeOnly
+
+trait ModifyMacroOptionSupport {
```
Comment (adamw) [inline core/src/main/scala-common/sttp.tapir/ModifyMacroOptionSupport.scala]: why `Option` in the name?

Comment (ghostbuster91) [reply core/src/main/scala-common/sttp.tapir/ModifyMacroOptionSupport.scala]: `Functor`?

Comment (ghostbuster91) [reply core/src/main/scala-common/sttp.tapir/ModifyMacroOptionSupport.scala]: `ModifyMacroCommon`?

File: core/src/main/scala-2.13+/sttp.tapir/ModifyMacroSupport.scala (hunk 2 above)
Comment (ghostbuster91) [reply core/src/main/scala-2.13+/sttp.tapir/ModifyMacroSupport.scala]: Thank you IntellijIdea, great job :1st_place_medal:

File: core/src/main/scala-common/sttp.tapir/ModifyMacroOptionSupport.scala (hunk 6 above)
Comment (adamw) [reply core/src/main/scala-common/sttp.tapir/ModifyMacroOptionSupport.scala]: functor sounds good :)

File: core/src/main/scala/sttp/tapir/Schema.scala (hunk 3 above)
Comment (ghostbuster91) [reply core/src/main/scala/sttp/tapir/Schema.scala]: > It seems it only causes a type parameter ot be required at use-site 
Only while using the unsafe variant, IMO that will be a very rare case.

**Comment by ghostbuster91** [comment]
We still miss the documentation, don't we? ;)

**Comment by adamw** [comment]
we do ;)
here: https://tapir-scala.readthedocs.io/en/latest/endpoint/customtypes.html#customising-derived-schemas

**Comment by adamw** [comment]
Releaed in 0.12.5, thanks! :)

**Pull request merged**